elasticsearch/_async/transport.py

```diff
@@ -0,0 +1,153 @@
+        """
+        Instantiate all the connections and create new connection pool to hold them.
+        Tries to identify unchanged hosts and re-use existing
+        :class:`~elasticsearch.Connection` instances.
+
+        :arg hosts: same as `__init__`
+        """
+        # construct the connections
+        def _create_connection(host):
+            # if this is not the initial setup look at the existing connection
+            # options and identify connections that haven't changed and can be
+            # kept around.
+            if hasattr(self, "connection_pool"):
+                for (connection, old_host) in self.connection_pool.connection_opts:
+                    if old_host == host:
+                        return connection
+
+            # previously unseen params, create new connection
+            kwargs = self.kwargs.copy()
+            kwargs.update(host)
+            return self.connection_class(**kwargs)
+
+        connections = map(_create_connection, hosts)
+
+        connections = list(zip(connections, hosts))
+        if len(connections) == 1:
+            self.connection_pool = DummyConnectionPool(connections)
+        else:
+            # pass the hosts dicts to the connection pool to optionally extract parameters from
+            self.connection_pool = self.connection_pool_class(
+                connections, **self.kwargs
+            )
+
+    def get_connection(self):
+        """
+        Retrieve a :class:`~elasticsearch.Connection` instance from the
+        :class:`~elasticsearch.ConnectionPool` instance.
+        """
+        return self.connection_pool.get_connection()
+
+    def mark_dead(self, connection):
+        """
+        Mark a connection as dead (failed) in the connection pool. If sniffing
+        on failure is enabled this will initiate the sniffing process.
+
+        :arg connection: instance of :class:`~elasticsearch.Connection` that failed
+        """
+        self.connection_pool.mark_dead(connection)
+
+    async def close(self):
+        if getattr(self, "sniffing_task", None):
+            self.sniffing_task.cancel()
+        await self.connection_pool.close()
+
+    async def perform_request(self, method, url, headers=None, params=None, body=None):
+        if body is not None:
+            body = self.serializer.dumps(body)
+
+            # some clients or environments don't support sending GET with body
+            if method in ("HEAD", "GET") and self.send_get_body_as != "GET":
+                # send it as post instead
+                if self.send_get_body_as == "POST":
+                    method = "POST"
+
+                # or as source parameter
+                elif self.send_get_body_as == "source":
+                    if params is None:
+                        params = {}
+                    params["source"] = body
+                    body = None
+
+        if body is not None:
+            try:
+                body = body.encode("utf-8")
```

Comment:
This should have the errors=surrogatepass flag since the real world is rather untidy. Without it, this introduces known issues.

https://github.com/elastic/elasticsearch-py-async/issues/62
https://github.com/elastic/elasticsearch-py-async/pull/69/files

Reply:
Agreed, will take from the sync version so it's identical.